File: deploy_to_vercel.py
#!/usr/bin/env python
"""
Vercel Deployment Script for GeoAI

This script facilitates the deployment of the GeoAI project to Vercel.
"""

import os
import sys
import subprocess
import json
from pathlib import Path

# Constants
VERCEL_CONFIG = "vercel.json"
FRONTEND_DIR = "frontend"
API_DIR = "api"
REQUIRED_ENV_VARS = ["VERCEL_TOKEN", "VERCEL_ORG_ID", "VERCEL_PROJECT_ID"]

def check_prerequisites():
    """Check that all prerequisites are met"""
    # Check if Vercel CLI is installed
    try:
        subprocess.run(["vercel", "--version"], 
                      check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        print("✅ Vercel CLI is installed")
    except (subprocess.SubprocessError, FileNotFoundError):
        print("❌ Vercel CLI is not installed. Please install it with: npm i -g vercel")
        return False
    
    # Check for required environment variables
    missing_vars = [var for var in REQUIRED_ENV_VARS if not os.environ.get(var)]
    if missing_vars:
        print(f"❌ Missing required environment variables: {', '.join(missing_vars)}")
        print("Please set them before deploying.")
        return False
    
    # Check if vercel.json exists
    if not os.path.exists(VERCEL_CONFIG):
        print(f"❌ {VERCEL_CONFIG} not found. Please create it first.")
        return False
    
    # Check if frontend package.json exists
    if not os.path.exists(os.path.join(FRONTEND_DIR, "package.json")):
        print(f"❌ {FRONTEND_DIR}/package.json not found.")
        return False
    
    print("✅ All prerequisites checked")
    return True

def prepare_api_for_vercel():
    """Prepare the API for Vercel deployment"""
    print("\nPreparing API for Vercel deployment...")
    
    # Create API directory if it doesn't exist
    os.makedirs(API_DIR, exist_ok=True)
    
    # Create requirements.txt for the API
    with open(os.path.join(API_DIR, "requirements.txt"), "w") as f:
        f.write("fastapi==0.104.0\n")
        f.write("uvicorn==0.24.0\n")
        f.write("pydantic==2.4.0\n")
        f.write("numpy==1.24.0\n")
        f.write("pandas==2.0.0\n")
        f.write("pillow==10.0.0\n")
        f.write("python-multipart==0.0.6\n")
        
    # Create a simple API file if it doesn't exist
    api_file = os.path.join(API_DIR, "index.py")
    if not os.path.exists(api_file):
        with open(api_file, "w") as f:
            f.write("""from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from typing import Dict, Any

app = FastAPI()

class HealthResponse(BaseModel):
    status: str
    version: str

@app.get("/api/health", response_model=HealthResponse)
async def health():
    """API health check endpoint"""
    return {
        "status": "healthy",
        "version": "1.0.0"
    }

@app.get("/api/regions")
async def get_regions():
    """Get available regions"""
    return {
        "regions": ["Alabama", "Mississippi", "Georgia", "Florida", "Tennessee"]
    }
""")
    print("✅ API prepared for Vercel")

def deploy_to_vercel():
    """Deploy the project to Vercel"""
    print("\nDeploying to Vercel...")
    
    # Run Vercel deployment
    try:
        result = subprocess.run(
            ["vercel", "--prod"], 
            check=True, 
            stdout=subprocess.PIPE, 
            stderr=subprocess.PIPE,
            text=True
        )
        print(result.stdout)
        print("✅ Deployment to Vercel completed successfully!")
        return True
    except subprocess.CalledProcessError as e:
        print(f"❌ Deployment failed: {e.stderr}")
        return False

def main():
    """Main function"""
    print("=" * 50)
    print("GeoAI Vercel Deployment")
    print("=" * 50)
    
    if not check_prerequisites():
        sys.exit(1)
    
    prepare_api_for_vercel()
    
    if deploy_to_vercel():
        print("\n" + "=" * 50)
        print("🎉 GeoAI is now deployed on Vercel!")
        print("=" * 50)
    else:
        print("\n" + "=" * 50)
        print("❌ Deployment failed. Please check the logs above.")
        print("=" * 50)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: deploy_to_vercel.js
node deploy_to_vercel.js --production#!/usr/bin/env node

/**
 * Vercel Deployment Script for GeoAI USA Agricultural Detection System
 * This script prepares and deploys the frontend to Vercel
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Configuration
const FRONTEND_DIR = path.join(__dirname, 'frontend');
const PROD_ENV_FILE = path.join(FRONTEND_DIR, '.env.production');

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  red: '\x1b[31m'
};

console.log(`${colors.blue}
========================================
   GeoAI Frontend Vercel Deployment
========================================
${colors.reset}`);

// Check if Vercel CLI is installed
function checkVercelCLI() {
  try {
    console.log(`${colors.yellow}Checking for Vercel CLI...${colors.reset}`);
    execSync('vercel --version', { stdio: 'ignore' });
    console.log(`${colors.green}✓ Vercel CLI is installed${colors.reset}`);
    return true;
  } catch (error) {
    console.log(`${colors.red}✗ Vercel CLI is not installed${colors.reset}`);
    console.log(`${colors.yellow}Installing Vercel CLI...${colors.reset}`);
    try {
      execSync('npm install -g vercel', { stdio: 'inherit' });
      console.log(`${colors.green}✓ Vercel CLI installed successfully${colors.reset}`);
      return true;
    } catch (installError) {
      console.error(`${colors.red}Error installing Vercel CLI: ${installError.message}${colors.reset}`);
      return false;
    }
  }
}

// Check and update environment variables
function updateEnvironmentVariables() {
  console.log(`${colors.yellow}Checking environment configuration...${colors.reset}`);
  
  if (!fs.existsSync(PROD_ENV_FILE)) {
    console.log(`${colors.red}✗ Production environment file not found at: ${PROD_ENV_FILE}${colors.reset}`);
    return false;
  }

  console.log(`${colors.green}✓ Environment configuration is ready${colors.reset}`);
  return true;
}

// Build the frontend
function buildFrontend() {
  console.log(`${colors.yellow}Building frontend for production...${colors.reset}`);
  try {
    process.chdir(FRONTEND_DIR);
    execSync('npm ci', { stdio: 'inherit' });
    execSync('npm run build', { stdio: 'inherit' });
    console.log(`${colors.green}✓ Frontend built successfully${colors.reset}`);
    return true;
  } catch (error) {
    console.error(`${colors.red}✗ Frontend build failed: ${error.message}${colors.reset}`);
    return false;
  }
}

// Deploy to Vercel
function deployToVercel(environment = 'production') {
  console.log(`${colors.yellow}Deploying to Vercel (${environment})...${colors.reset}`);
  try {
    process.chdir(FRONTEND_DIR);
    const command = environment === 'production' 
      ? 'vercel --prod'
      : 'vercel';
    
    execSync(command, { stdio: 'inherit' });
    console.log(`${colors.green}✓ Deployment successful${colors.reset}`);
    return true;
  } catch (error) {
    console.error(`${colors.red}✗ Deployment failed: ${error.message}${colors.reset}`);
    return false;
  }
}

// Main function
async function main() {
  const args = process.argv.slice(2);
  const isProd = args.includes('--prod') || args.includes('--production');
  const environment = isProd ? 'production' : 'preview';
  
  console.log(`${colors.blue}Deploying to ${environment} environment${colors.reset}`);
  
  // Run deployment steps
  const hasVercel = checkVercelCLI();
  if (!hasVercel) {
    process.exit(1);
  }
  
  const envReady = updateEnvironmentVariables();
  if (!envReady) {
    process.exit(1);
  }
  
  const buildSuccess = buildFrontend();
  if (!buildSuccess) {
    process.exit(1);
  }
  
  const deploySuccess = deployToVercel(environment);
  if (!deploySuccess) {
    process.exit(1);
  }
  
  console.log(`${colors.blue}
========================================
   Deployment Complete! 
   
   Your GeoAI frontend is now live on Vercel.
   Check your Vercel dashboard for the URL.
========================================
${colors.reset}`);
}

// Run the script
main().catch(error => {
  console.error(`${colors.red}Unexpected error: ${error.message}${colors.reset}`);
  process.exit(1);
});